File: src/leetCodeProblems/DFSUsingRecursion/BinaryTree/BinaryTreeIsSymmetric101.java
package leetCodeProblems.BinaryTree;

/**
 * LeetCode - https://leetcode.com/problems/symmetric-tree/
 */
public class BinaryTreeIsSymmetric {

    static class TreeNode {
        int val;

        TreeNode left;

        TreeNode right;

        TreeNode(int val) {
            this.val = val;
        }
    }

    private boolean isSymmetricUtil(TreeNode node, TreeNode symmetricNode) {

        if (node == null && symmetricNode == null) {
            return true;
        }

        if (node == null || symmetricNode == null || node.val != symmetricNode.val) {
            return false;
        }

        boolean isLeftChildSymmetric = isSymmetricUtil(node.left, symmetricNode.right);
        boolean isRightChildSymmetric = isSymmetricUtil(node.right, symmetricNode.left);

        if (!isLeftChildSymmetric || !isRightChildSymmetric) {
            return false;
        }

        return true;
    }

    public boolean isSymmetric(TreeNode root) {

        if ( root == null ) {
            return true;
        }

        return isSymmetricUtil(root.left, root.right);
    }

    public static void main(String[] args) {

        /*TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(2);
        root.left.left = new TreeNode(3);
        root.left.right = new TreeNode(4);
        root.right.left = new TreeNode(4);
        root.right.right = new TreeNode(3); // Expected O/P = true*/

        /*TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.left.right = new TreeNode(3);
        root.right = new TreeNode(2);
        root.right.right = new TreeNode(3); // Expected O/P = false*/

        TreeNode root = new TreeNode(1);
        root.right = new TreeNode(2); // Expected O/P = false

        BinaryTreeIsSymmetric obj = new BinaryTreeIsSymmetric();
        System.out.println(obj.isSymmetric(root));

    }
}

File: src/leetCodeProblems/DFSUsingRecursion/BinaryTree/BinaryTreeIsSymmetric101.go
package main

/*
- Leetcode - https://leetcode.com/problems/symmetric-tree/
- Time - O(n)
- Space - O(1)
*/

func isSymmetricUtil(node1 *TreeNode, node2 *TreeNode) bool {

	if node1 == nil && node2 == nil {
		return true
	}

	if node1 == nil || node2 == nil {
		return false
	}

	if node1.Val == node2.Val &&
		isSymmetricUtil(node1.Left, node2.Right) &&
		isSymmetricUtil(node1.Right, node2.Left) {
		return true
	}

	return false
}

func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}

	return isSymmetricUtil(root.Left, root.Right)
}

// func main() {
// 	// root := new(TreeNode) //Returns pointer to TreeNode object
// 	// root.Val = 1
// 	// root.Left = new(TreeNode)
// 	// root.Left.Val = 2
// 	// root.Left.Left = new(TreeNode)
// 	// root.Left.Left.Val = 3
// 	// root.Left.Right = new(TreeNode)
// 	// root.Left.Right.Val = 4

// 	// root.Right = new(TreeNode)
// 	// root.Right.Val = 2
// 	// root.Right.Left = new(TreeNode)
// 	// root.Right.Left.Val = 4
// 	// root.Right.Right = new(TreeNode)
// 	// root.Right.Right.Val = 3

// 	// log.Println(isSymmetric(root))

// 	root := new(TreeNode) //Returns pointer to TreeNode object
// 	root.Val = 1
// 	root.Left = new(TreeNode)
// 	root.Left.Val = 2
// 	root.Left.Right = new(TreeNode)
// 	root.Left.Right.Val = 3

// 	root.Right = new(TreeNode)
// 	root.Right.Val = 2
// 	root.Right.Right = new(TreeNode)
// 	root.Right.Right.Val = 3

// 	log.Println(isSymmetric(root))

// }
